internal/cmd/temporal-features-test-proxy/main.go

```diff
@@ -0,0 +1,752 @@
+	"io"
+	"io/fs"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"path"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/sdk/client"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/status"
+)
+
+const HelpText = `The test proxy exposes the following control endpoints:
+
+- POST /quit
+  Shut down the proxy and exit.
+
+- POST /restart
+  Gracefully shut down the gRPC server, then start it again.
+
+  - Query param: sleep=<duration>
+    Forces the restart to block for the given duration; default: 0s.
+
+  - Query param: forceful=<bool>
+    If true, forces a non-graceful shutdown; default: false.
+
+- POST /reject
+  Immediately reject incoming gRPC requests with UNAVAILABLE.
+
+- POST /accept
+  Accept incoming gRPC requests; this is the default.
+
+- POST /freeze
+  Block on incoming accepted gRPC requests.
+
+- POST /thaw
+  Process incoming accepted gRPC requests immediately; this is the default.
+`
+
+var ErrUnknownCommand = errors.New("unknown command")
+
+var (
+	flagTrace   bool
+	flagControl string
+	flagListen  string
+	flagDial    string
+
+	gListenConfig  net.ListenConfig
+	gExitCh        chan struct{}
+	gRootContext   context.Context
+	gServerMutex   sync.Mutex
+	gControlServer ControlServer
+	gProxyServer   ProxyServer
+
+	gStateMutex     sync.Mutex
+	gStateCond      *sync.Cond
+	gStateRejecting bool
+	gStateFrozen    bool
+)
+
+func init() {
+	flag.BoolVar(&flagTrace, "trace", false, "enable tracing logs")
+	flag.StringVar(&flagControl, "control", "", "TCP host:port to listen on for HTTP control commands")
+	flag.StringVar(&flagListen, "listen", "", "TCP host:port to listen on for proxying to -dial")
+	flag.StringVar(&flagDial, "dial", "", "TCP host:port to connect to")
+}
+
+func main() {
+	flag.Parse()
+
+	if flagControl == "" {
+		Fatal(1, "must specify -control")
+		panic(nil)
+	}
+	if flagListen == "" {
+		Fatal(1, "must specify -listen")
+		panic(nil)
+	}
+	if flagDial == "" {
+		Fatal(1, "must specify -dial")
+		panic(nil)
+	}
+
+	gExitCh = make(chan struct{})
+	gStateCond = sync.NewCond(&gStateMutex)
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+	ctx, stop := signal.NotifyContext(ctx, os.Interrupt)
+	defer stop()
+	gRootContext = ctx
+
+	gControlServer.Init(flagControl)
+	err := gControlServer.Run(ctx)
+	if err != nil {
+		Fatal(2, "%v", err)
+		panic(nil)
+	}
+
+	needControlClose := true
+	defer func() {
+		if needControlClose {
+			gControlServer.ForceClose()
+		}
+	}()
+
+	gProxyServer.Init(flagListen, flagDial)
+	err = gProxyServer.Run(ctx)
+	if err != nil {
+		Fatal(2, "%v", err)
+		panic(nil)
+	}
+
+	needProxyClose := true
+	defer func() {
+		if needProxyClose {
+			gProxyServer.ForceClose()
+		}
+	}()
+
+	Info("HTTP control server is running on: %s", flagControl)
```

Comment:
Nit: If user passes `127.0.0.1:0` this would show 0 rather than the bound port, yes? 

Not really a problem since we're predetermining available ports and using those, but thought I'd mention it since this can be run directly in theory.

Reply:
When this is made a library as part of a runner, we should let this listener pick the random port and extract the address out of it instead temporary-listener-free-port approach (which is a good approach in general, but won't be needed when this is not a separate process).